internals/cli/env.go

```diff
@@ -165,47 +158,211 @@ func (a *App) CheckStrictEnv() error {
 
 // CommandClause represents a command clause in a command0-line application.
 type CommandClause struct {
-	*kingpin.CmdClause
+	*cobra.Command
 
 	name string
-	app  *App
+	App  *App
+}
+
+func (cmd *CommandClause) BoolVarP(reference *bool, name, shorthand string, def bool, usage string, hasEnv bool, persistent bool) {
+	if persistent {
+		cmd.PersistentFlags().BoolVarP(reference, name, shorthand, def, usage)
+	} else {
+		cmd.Flags().BoolVarP(reference, name, shorthand, def, usage)
+	}
+
+	if hasEnv {
+		cmd.FlagRegister(name, usage)
+	}
+}
```

Comment:
We should try to reduce code duplication here while still giving the user of the package the option to request an environment variable for a given flag.

The best alternative I could come up with is the following:
We switch the syntax for registering flags from this:
```go
cmd.BoolVarP(ref, name, short, def, usage, true, false)
```
to this:
```go
cmd.WithEnvVar().Flags().BoolVarP(ref, name, short, def, usage)
```
and from this:
```go
cmd.BoolVarP(ref, name, short, def, usage, false, true)
```
to this:
```go
cmd.PersistentFlags().BoolVarP(ref, name, short, def, usage)
```

This way we only have to implement one method: `cmd.WithEnvVar()`
The other two methods on `cmd` (`cmd.Flags()` and `cmd.PersistentFlags()`) won't have to be implement since the `CommandClause` embeds the `*cobra.Command` type, which already implements these.

The only issue with this idea is that the `WithEnvVar()` method cannot set the default value of the flag to the value of the corresponding environment variable right away, since it runs before the flag is registered. To circumvent this issue we could make the `WithEnvVar()` method just store the command and flag name somewhere, and later, after all flags and commands are registered we could call a function that sets the default values for all flags registered with `WithEnvVar()` to their corresponding envrionment variable's value.

Let me know if you can think of other alternatives. Otherwise, I suggest we try implementing this solution and see if it raises any unforeseen difficulties.

Reply:
Also let me know if my explanation isn't clear enough or if you have any related questions.

Reply:
I like what you're proposing. I will look into it first thing tomorrow.

Reply:
Saw this pass by. If you haven't already, check out how kingpin does it: `ArgClause.Envar()` and `FlagClause.Envar()`. Could be inspiration. https://godoc.org/github.com/alecthomas/kingpin

Also, consider that in the current CLI all flags by default have an environment variable. I think we will want to keep that, instead of having to specifically call `.WithEnvVar()` for every flag. That way we can easily forget to make that call and introduce unexpected behavior. 

Small considerations, it looks like you're on top of it. Just thought I'd share my two cents.

Reply:
> Saw this pass by. If you haven't already, check out how kingpin does it: `ArgClause.Envar()` and `FlagClause.Envar()`. Could be inspiration. https://godoc.org/github.com/alecthomas/kingpin
> 
> Also, consider that in the current CLI all flags by default have an environment variable. I think we will want to keep that, instead of having to specifically call `.WithEnvVar()` for every flag. That way we can easily forget to make that call and introduce unexpected behavior.
> 
> Small considerations, it looks like you're on top of it. Just thought I'd share my two cents.

Thank you for your suggestion!
This, indeed, is the behaviour we strived for, i.e. making as hard as possible to omit the creation of an env var for a flag. 
At the moment, specifying that a certain flag has an environment variable only takes a `true` passed as argument to the flag registration function. See the function signature for a clear example:
```go
func (cmd *CommandClause) BoolVarP(reference *bool, name, shorthand string, def bool, usage string, hasEnv bool, persistent bool) 
```
This design was decided together with @SimonBarendse, who cleverly remarked that overriding the functions that cobra provides in order to provide functionality for environment variables would take less code than to actually call a function on every flag that's initialized. However, as it can be seen, this by itself brought quite some code duplication: 12 functions (with about 10 lines/function), so far, and if we implement @Marton6 's proposal to use cobra's default StringToString implementation, it's going to be 13 or 14.
Therefore, this still needs to be debated, since it's one of the most important design decisions that is to be made here. We'll look more into it tomorrow.

Reply:
After a short discussion and a bit of experimentation it seems that my idea does not eliminate the need for writing a function/flag type, thus we decided not to implement it. We are still looking for alternatives.

Reply:
Can this conversation be resolved?

Reply:
Yes. Thank you for implementing it.